### PR TITLE
API: Introduce ``copy`` argument for ``np.asarray`` [Array API]

### PR DESCRIPTION
**Pull request by mtsokol**
Tracking issue: https://github.com/numpy/numpy/issues/25076

Hi @seberg,

Here's a PR with an initial work on introducing `copy` keyword to `np.asarray`.

I started looking at the internals that we discussed during the triage meeting, but first, I wanted to try with the simplest approach: The implementation of `np.asarray` is mostly the same as `np.array` - it just skips a few parameters. One of them is `copy` which is hardcoded to `False` in `np.asarray`.

Here I added `copy` with a default value `False`, so it's a backward compatible change. For `copy=True` it behaves the same way as `np.array(..., copy=True)`. To sum up - the `copy` mechanism didn't change here, only the `copy` keyword is added for compatibility. 

WDYT?

### CODE REVIEW

**Review by seberg — COMMENTED** [review]
Not sure what you are looking for.  If we need to add the `copy` kwarg to `asarray`, this is the right way.

The first interesting part is vetting the "no copy" logic (I don't remember if it is bad, but I definitely don't trust it without someone auditing it once more).
The other interesting part is places where `copy=False` already exists.

For `np.array()` that is likely fine (a brief search may be nice), some people *will* be using `np.array(..., copy=False)` or actually `np.array(x, None, False)` (or a dtype) [^1].
The hope for those use-cases is that most of them will actually like the new meaning, and if they don't they will fail gracefully (it would be cute to mention `copy=None` and `asarray()` in the error message).

The trickier thing is probably `arr.astype(original, copy=False)`.  The problem is that this is very much *very* common code and such code will definitely expect the copy to be necessary often.
Also there, you can kust replace it with `copy=None`, but I am worried about the amount of churn we might create, since `copy=False` was always a best practice!
(It may be that while a best-practice in many places, it isn't nearly as common in end-user code, but I suspect it is fairly common.)

[^1]: That may look crazy, but the the use of kwargs used to come with a large overhead and `asarray` was a Python wrapper using kwargs making it very slow for just returning the original object.

File: numpy/_core/src/multiarray/multiarraymodule.c [hunk 1]
```diff
@@ -1752,6 +1752,7 @@ array_asarray(PyObject *NPY_UNUSED(ignored),
         PyObject *const *args, Py_ssize_t len_args, PyObject *kwnames)
 {
     PyObject *op;
+    _PyArray_CopyMode copy = NPY_COPY_IF_NEEDED;
```
[the remark this answers is not in the document]
Comment (mtsokol) [reply numpy/_core/src/multiarray/multiarraymodule.c]: Sure! Renamed to `NPY_COPYMODE`.

File: numpy/_core/src/multiarray/multiarraymodule.c [hunk 2]
```diff
@@ -1779,11 +1781,12 @@ array_asarray(PyObject *NPY_UNUSED(ignored),
         }
     }
     else {
+        /* Fast path for symmetry (we copy by default which is slow) */
```
[the remark this answers is not in the document]
Comment (mtsokol) [reply numpy/_core/src/multiarray/multiarraymodule.c]: Right, I removed it.

**Comment by ngoldbaum** [comment]
It looks like there are a number of places pandas uses `arr.astype(..., copy=False)` (although [lots of other code does it too](https://github.com/search?q=%22arr.astype%22+%22copy%3DFalse%22&type=code)). Maybe running the pandas test suite against a numpy with this PR applied will be enough to discover whether that causes breakages?

**Comment by ngoldbaum** [comment]
You can use https://pandas-coverage-12d2130077bc.herokuapp.com to find which pandas tests hit a particular line of code in the pandas codebase.

**Comment by seberg** [comment]
It seems to me that there is a worm somewhere in all of this.  What is `astype()` supposed to do?  The documentation says that `False` allows a copy if needed.  That is also true for the docs in the array-api.
But doesn't this change it so that `False` enforces something stricter at least sometimes?

So, is the meaning of `copy=False` just different for different functions?  That solves a lot of problems, although it is also a bit confusing...

**Comment by seberg** [comment]
I am trying to avoid to think about it...  I think it is strange to diverge the meaning, but breaking `astype(copy=False)` is a huge amount of churn that needs a bit clearer motivation.

**Comment by mtsokol** [comment]
Hi @seberg,

I updated the PR according to our discussion during the triage call and https://github.com/data-apis/array-api/issues/721. 

The new semantics `None/True/False` are now present only in `np.array` and `np.asarray`, and default values didn't change. `np.astype` and `ndarray.astype` `copy` behavior also stayed the same.

In the C-level code there is an updated definition of `NPY_COPYMODE` to adhere to the new semantics and a new `NPY_ASTYPECOPYMODE` enum for `astype` to keep the same behavior.

**Comment by seberg** [comment]
I do *not* see it as my decision to give the OK on changing how copy behaves in `asarray`, `array`, and friends.  If there isn't a consensus, it should be in the NEP to propose to change it.

But, I do not like the change.  The reason why I can live with the redefinition of `copy=False` in `np.array(..., copy=False)` is that it can be (and often is) spelled as `np.asarray(...)` (this is assuming downstream churn seems acceptable, I have no idea how bad it is).  So, use `asarray`, not `copy=None`.
That is: `copy=None` is acceptable maybe, but partially because we at least normally do not use it explicitly (not sure about `meshgrid`, which defaults to `copy=True`).

**Comment by seberg** [comment]
In fact, we could just split out the `array(..., copy=False)` -> `asarray` refactor to make this PR smaller.

**Comment by mtsokol** [comment]
> In fact, we could just split out the `array(..., copy=False)` -> `asarray` refactor to make this PR smaller.

Sure, I updated `array` to `asarray` wherever it was applicable.

**Comment by ngoldbaum** [comment]
There are conflicts now that I've merged the other array API PRs. I'm also holding off on merging this because the discussion around this in the NEP PR hasn't converged.

File: numpy/_core/src/multiarray/conversion_utils.c [hunk 3]
```diff
@@ -258,7 +257,32 @@ PyArray_CopyConverter(PyObject *obj, _PyArray_CopyMode *copymode) {
...
+    }
+    else {
+        npy_bool bool_copymode;
+        if (!PyArray_BoolConverter(obj, &bool_copymode)) {
+            return NPY_FAIL;
+        }
+        int_copymode = (int)bool_copymode;
+    }
+
+    *copymode = (NPY_ASTYPECOPYMODE)int_copymode;
```
Comment (ngoldbaum) [inline numpy/_core/src/multiarray/conversion_utils.c:285]: We talked about this at the meeting and there was a desire to also support and document the strings `'always'`, `'never'` and `'if-needed'`.

**Comment by rgommers** [comment]
A few high-level notes:
- The PRs by @mtsokol to adapt SciPy and scikit-learn for the changes in this PR (https://github.com/scipy/scipy/pull/19997, https://github.com/scikit-learn/scikit-learn/pull/28323) are pretty small, so that's a good sign.
- I updated the description in the NEP (gh-25542) with more detail. I think the key change not yet reflected in this PR is to add a `copy=None` keyword to `__array__`. I imagine it can be used like so:

```python
if copy is not None:
    try:
        obj = obj.__array__(dtype, copy=copy)
    except TypeError:
        # foreign __array__ implementation doesn't yet have copy=None in signature
        obj = obj.__array__(dtype)
        _msg = f"`{type(obj)}.__array__` is missing a `copy` keyword"
        warnings.warn(_msg, DeprecationWarning, stacklevel=2)
```

**Comment by seberg** [comment]
> The last commit that I pushed adds copy keyword argument to ndarray.__array__ with tests.

Thanks.  It adds it to `ndarray.__array__`, which is needed but not the interesting part.  The interesting part is to try to pass it where we *call* `__array__` (probably only when `copy` is indicated).

**Comment by mtsokol** [comment]
> The interesting part is to try to pass it where we call `__array__` (probably only when copy is indicated).

@seberg The only relevant place that I found where `__array__` gets called  was inside `np.array` implementation. I modified it in such a way that `copy=...` argument from `np.array(...)` is passed to `__array__`. If `__array__` doesn't support it a warning is raised and it gets called again without `copy`.

Classes in tests/codebase that implement `__array__` function now have `dtype` and `copy` arguments to adhere to the protocol - tomorrow I will update the documentation to cover this change.

**Comment by mtsokol** [comment]
Ok, I updated docs so that `__array__` part mentions `copy` argument.

File: numpy/_core/src/multiarray/methods.c [hunk 4]
```diff
@@ -967,13 +963,26 @@ array_getarray(PyArrayObject *self, PyObject *args)
...
+        if (newtype == NULL || PyArray_EquivTypes(PyArray_DESCR(self), newtype)) {
+            return (PyObject *)self;
+        }
+        if (copy == NPY_COPY_IF_NEEDED) {
+            ret = PyArray_CastToType(self, newtype, 0);
+            Py_DECREF(self);
+            return ret;
+        } else { // copy == NPY_COPY_NEVER
+            PyErr_SetString(PyExc_ValueError,
+                            "Unable to avoid copy while creating an array from given array.");
```
[the remark this answers is not in the document]
Comment (mtsokol) [reply numpy/_core/src/multiarray/methods.c:983]: Right, there's a `Py_INCREF(self)` before, and in this path we set an error instead of returning `self`. I added `Py_DECREF(self)` here.

File: numpy/_core/src/multiarray/ctors.c [hunk 5]
```diff
@@ -2458,12 +2439,30 @@ PyArray_FromArrayAttr_int(
...
+        PyErr_Fetch(&type, &value, &traceback);
+        if (PyUnicode_Check(value) && PyUnicode_CompareWithASCIIString(value,
+                    "__array__() got an unexpected keyword argument 'copy'") == 0) {
+            if (PyErr_WarnEx(PyExc_UserWarning,
+                             "__array__ should implement 'dtype' and 'copy' keywords", 1) < 0) {
+                return NULL;
+            }
+            new = PyObject_Call(array_meth, args, NULL);
+        } else {
+            PyErr_Restore(type, value, traceback);
```
[the remark this answers is not in the document]
Comment (ngoldbaum) [reply numpy/_core/src/multiarray/ctors.c:2464]: Thanks for double-checking!

File: numpy/_core/src/multiarray/ctors.c [hunk 6]
```diff
@@ -2458,12 +2439,30 @@ PyArray_FromArrayAttr_int(
...
+    if (PyErr_Occurred()) {
+        PyObject *type, *value, *traceback;
+        PyErr_Fetch(&type, &value, &traceback);
+        if (PyUnicode_Check(value) && PyUnicode_CompareWithASCIIString(value,
+                    "__array__() got an unexpected keyword argument 'copy'") == 0) {
+            if (PyErr_WarnEx(PyExc_UserWarning,
+                             "__array__ should implement 'dtype' and 'copy' keywords", 1) < 0) {
+                return NULL;
+            }
+            new = PyObject_Call(array_meth, args, NULL);
```
[the remark this answers is not in the document]
Comment (ngoldbaum) [reply numpy/_core/src/multiarray/ctors.c:2462]: There's still an issue. You created `new` above, but are rebinding it here to a new value, which leaks the old version of `new`.

I would replace this with:

```suggestion
            Py_SETREF(new, PyObject_Call(array_meth, args, NULL));
```

File: doc/source/reference/c-api/array.rst [hunk 7]
```diff
@@ -515,7 +515,8 @@ From other objects
 
     Return an ndarray object from a Python object that exposes the
     :obj:`~numpy.class.__array__` method. The :obj:`~numpy.class.__array__`
-    method can take 0, or 1 argument ``([dtype])``. ``context`` is unused.
+    method can take 0, 1 or 2 arguments ``(dtype=None, copy=None)``.
```
Comment (ngoldbaum) [inline doc/source/reference/c-api/array.rst:518]: Sorry for not flagging this on the first review, can you rephrase this to emphasize that third-party `__array__` implementations *must* take `dtype` and `copy` keyword arguments? Can you also document the meaning of those keyword arguments and explain that third-party code or numpy might break if `__array__` ignores the keywords.

**Review by ngoldbaum — COMMENTED** [review]
Sorry for not following up quickly after you responded to my last review. I think there's still one refcounting issue. I also noticed a separate more minor issue in the docs but I think is worth addressing to reduce confusion from users.

@seberg if you have a chance, it would be nice if you could confirm you're ok with the semantics in this version of the PR just so I can make sure we're all on the same page.

**Comment by seberg** [comment]
I am happy with the addition.  Note that it will be quite noisy unless you do not pass copy=None and take it this very slow.

Code wise happy if you are.  matching error exact is likely nice.  Passing positionally for simplicity woukd be fine to me or using vectorcall would be best (much quicker probably and probably also simpler).

File: numpy/_core/src/multiarray/ctors.c (hunk 6 above)
[the remark this answers is not in the document]
Comment (mtsokol) [reply numpy/_core/src/multiarray/ctors.c]: Done!

File: doc/source/reference/c-api/array.rst (hunk 7 above)
Comment (mtsokol) [reply doc/source/reference/c-api/array.rst]: Ok! I mentioned there that `dtype` and `copy` are mandatory. In `~numpy.class.__array__` docs (which is linked here) I repeated it and explained the meaning of these arguments.

**Comment by ngoldbaum** [comment]
Merging this. Thanks @mtsokol! Also thanks for getting all the NEP 56 PRs across the line! 🎆 🥳 

If you are reading this because this PR broke your nightly tests, take a look at the new docs, the NEP, and the scipy and scikit-learn PRs. If there is some sort of breakage in your code that is not straightforward to deal with in a backward compatible way, please open an issue.

**Comment by rgommers** [comment]
Thanks for the hard work on this! Still a few things to iron out I think, xref gh-25916.
